[로또] 코나스 미션 제출합니다.

Conatuseus opened the pull request:
생각을 너무 많이 하다보니 코드가 더 꼬여버린 것 같습니다.
부족하지만 많은 피드백 부탁드립니다.

Conatuseus committed:
docs : README 작성

Squashed commit of the following:

commit 8fa116350ea3a973842a384509ee118b3af2fc38
Author: conatuseus <conatuseus@gmail.com>
Date:   Tue May 28 14:47:19 2019 +0900

    README 작성

Conatuseus committed:
feature : 로또 번호 생성하는 generator 구현

LottoNumberGenerator 테스트 코드 및 클래스 구현
README 수정

Squashed commit of the following:

commit 9d54d5b7c25d6f9a0b2e66fbe9e7eb5f843b0511
Author: conatuseus <conatuseus@gmail.com>
Date:   Tue May 28 14:58:45 2019 +0900

    LottoNumberGenerator 테스트 코드 작성 및 구현

Conatuseus committed:
feature : 로또 번호 테스트 코드 추가 및 구현

LottoNumber class 구현
lottoNumber가 올바르지 않으면 예외처리

Squashed commit of the following:

commit c24352c7b30845b423d6dc2bee3ad877a1934712
Author: conatuseus <conatuseus@gmail.com>
Date:   Tue May 28 16:06:17 2019 +0900

    Lotto 자동 생성을 위한 코드 주석처리

commit 74d9751ff91b4ebf3532972a1965ad82e07f38a6
Author: conatuseus <conatuseus@gmail.com>
Date:   Tue May 28 15:45:55 2019 +0900

    LottoNumber class 구현

Conatuseus committed:
feature : 로또 자동 생성 클래스 구현

로또 수를 random으로 6개 만들던 것을
lotto 수 46개를 저장해서 섞어서 6개를 가져오는 클래스로 변경

Squashed commit of the following:

commit 8be9a06db29acaf92d9af88128a6d9fc6ea0428e
Author: conatuseus <conatuseus@gmail.com>
Date:   Tue May 28 16:31:13 2019 +0900

    LottoNumberGenerator 삭제. LottoGenerator에서 자동로또 생성하도록하기 위해

commit 3c154e43fc26407a6c0960c742cdad355ed750be
Author: conatuseus <conatuseus@gmail.com>
Date:   Tue May 28 16:30:21 2019 +0900

    불필요한 주석 제거

commit 9465e2f709e3c3dfb90371e23fc97e3ef341902f
Author: conatuseus <conatuseus@gmail.com>
Date:   Tue May 28 16:29:21 2019 +0900

    LottoGenerator 구현

Conatuseus committed:
feature : 로또 클래스 구현

Lotto class 구현
LottoTest 구현
로또 자동생성 클래스 수정
해당 테스트코드 수정

Squashed commit of the following:

commit 5af9b6d7aa1b51740a13e6ef57712e22e7182cf1
Author: conatuseus <conatuseus@gmail.com>
Date:   Wed May 29 11:44:11 2019 +0900

    LottoAutoGenerator 수정

commit 40e0ed39818ae0b9cb4896ef6459753a588adacc
Author: conatuseus <conatuseus@gmail.com>
Date:   Wed May 29 10:50:47 2019 +0900

    Lotto Test code 수정, LottoAutoGenerator 수정

commit efee64cc4503fca1c1eee5a9159023c49fc8fdf1
Author: conatuseus <conatuseus@gmail.com>
Date:   Tue May 28 17:43:21 2019 +0900

    주석처리 코드 삭제

commit 78c0fe180914be34608ccbbab6c8ce0824c088ee
Author: conatuseus <conatuseus@gmail.com>
Date:   Tue May 28 17:42:39 2019 +0900

    LottoGenerator -> LottoAutoGenerator 클래스명 변경

commit 45dcdb38e67c0ed61b9cb59dfba76e556e120b14
Author: conatuseus <conatuseus@gmail.com>
Date:   Tue May 28 17:01:21 2019 +0900

    TestCode 구현

commit a6ddb59f498f7736764bc44705f8905fb26dbcb6
Author: conatuseus <conatuseus@gmail.com>
Date:   Tue May 28 16:54:12 2019 +0900

    Lotto class 구현

Conatuseus committed:
feature : 지난주 당첨 로또 구현

지난주 당첨 로또인 Winning Lotto 구현
Lottos -> UserLottos 로 이름 변경

Squashed commit of the following:

commit 0075d80fe47ca9b00137c341f3b112f1223a96d3
Author: conatuseus <conatuseus@gmail.com>
Date:   Thu May 30 14:53:01 2019 +0900

    불필요한 코드 제거

commit 5a89edb4978faf6079458a3345b64c44e1e87428
Author: conatuseus <conatuseus@gmail.com>
Date:   Thu May 30 14:52:33 2019 +0900

    불필요한 코드 제거

commit 4e785e0e82115b36e310b125ed876d9b195f01e2
Author: conatuseus <conatuseus@gmail.com>
Date:   Thu May 30 14:50:40 2019 +0900

    WinningLotto 구현

commit 36183158c24860e84dd54c3468ec98a657d56824
Author: conatuseus <conatuseus@gmail.com>
Date:   Wed May 29 11:52:05 2019 +0900

    로또 번호 1-45으로 수정 (기존 : 1~46)

commit 3f5f3096d101c5fd085860896309185450f7eb91
Author: conatuseus <conatuseus@gmail.com>
Date:   Wed May 29 11:50:50 2019 +0900

    README 수정

Conatuseus committed:
feature : User 구현

User 클래스 구현
구입 금액 객체화
테스트 코드 작성

Squashed commit of the following:

commit 7a0be6ff09a21eb0dd444fe3d4b96c6f58879fb8
Author: conatuseus <conatuseus@gmail.com>
Date:   Thu May 30 16:09:01 2019 +0900

    Money 객체화, User 구현 및 테스트코드 작성

Conatuseus committed:
feature : InputView 구현, WinningLottoParser 구현

InputView 구현
WinningLottoParser 구현
Test 코드 추가

Squashed commit of the following:

commit 3c13f96350ddc6a2536127368d58663453142a28
Author: conatuseus <conatuseus@gmail.com>
Date:   Thu May 30 17:22:00 2019 +0900

    WinningLottoParseTest 구현, 인덱스로 LottoNumber가져오는 메소드 구현

commit f347cf59f3e7d27b6e515978686558311cd0bf75
Author: conatuseus <conatuseus@gmail.com>
Date:   Thu May 30 17:21:44 2019 +0900

    WinningLottoParseTest 구현, 인덱스로 LottoNumber가져오는 메소드 구현

commit a96d1c24772c48814f0900da98e9bf50f410e516
Author: conatuseus <conatuseus@gmail.com>
Date:   Thu May 30 17:10:08 2019 +0900

    Money Test, WinningLottoTest 구현

commit 533958090a198729d01f5661b01004f6c12ebb42
Author: conatuseus <conatuseus@gmail.com>
Date:   Thu May 30 17:02:43 2019 +0900

    InputView, WinningLottoPaser 구현

Conatuseus committed:
feature : 입출력 기능 추가

돈 입출력 기능 추가
자동생성된 로또 출력 기능 추가
당첨 로또 입력 기능 추가

Squashed commit of the following:

commit dfced86a5c7f988fbfd70344fef07bbb21d4e835
Author: conatuseus <conatuseus@gmail.com>
Date:   Thu May 30 18:03:31 2019 +0900

    돈 입출력, 자동생성된 로또 출력, 당첨로또 입력 기능 추가

commit dd3e8a4f722fcb9d3fc15ef8b142da45753b0b88
Author: conatuseus <conatuseus@gmail.com>
Date:   Thu May 30 17:28:45 2019 +0900

    README 수정

commit e1c47b8bf7359997872fee1b61619a4c28c8ca3e
Author: conatuseus <conatuseus@gmail.com>
Date:   Thu May 30 17:26:29 2019 +0900

    README 수정

Conatuseus committed:
feature : enum rank 구현

Conatuseus committed:
feature : 당첨 통계 기능, 출력 기능 구현

당첨 통계 기능 추가
출력 기능 추가
코드 정리

Squashed commit of the following:

commit fa76576e6f8d3cc720a0c335c2f018ffbd8772f8
Author: conatuseus <conatuseus@gmail.com>
Date:   Mon Jun 3 16:01:50 2019 +0900

    코드 정리

commit b97dbe563b60e7f303e6a5d74b9129b364f9a1dd
Author: conatuseus <conatuseus@gmail.com>
Date:   Mon Jun 3 15:46:19 2019 +0900

    당첨 통계 구하는 기능, 출력 기능 추가

commit 69cf720b4c94c84f75cafe2af1224b55e10f3412
Author: conatuseus <conatuseus@gmail.com>
Date:   Mon Jun 3 14:17:03 2019 +0900

    수익률 구하는 매서드 추가

Conatuseus committed:
refactor : 상수화, 테스트 코드 수정

Squashed commit of the following:

commit 1e07e2cd135c5d11000539b23c868538e1f8a4fd
Author: conatuseus <conatuseus@gmail.com>
Date:   Mon Jun 3 21:47:36 2019 +0900

    상수화 및 테스트코드 수정

Conatuseus committed:
docs : README 작성

Squashed commit of the following:

commit 08e11f163e77da35a92e8c6c859d131a2c0ead64
Author: conatuseus <conatuseus@gmail.com>
Date:   Mon Jun 3 21:51:01 2019 +0900

    Todo 작성

Conatuseus committed:
feature : 수동 로또 수 입력 기능 구현

Squashed commit of the following:

commit f28e725cf50f5ccc8aa28d520622419a17909a1f
Author: conatuseus <conatuseus@gmail.com>
Date:   Tue Jun 4 14:40:29 2019 +0900

    수동 구매 로또 수 입력 기능 추가

Conatuseus committed:
feature : 로또 수동 입력 기능 구현

Squashed commit of the following:

commit 705737fba2dcff5c78c4d65b556bee569213b6f0
Author: conatuseus <conatuseus@gmail.com>
Date:   Tue Jun 4 16:21:53 2019 +0900

    코드 정리

commit 091b4ffa51f8d34f7733777b443a5944b9345cae
Author: conatuseus <conatuseus@gmail.com>
Date:   Tue Jun 4 16:06:41 2019 +0900

    로또 번호 수동 입력 기능 구현

pobiconan commented on src/main/java/lotto/Main.java:
```diff
@@ -0,0 +1,33 @@
+
+public class Main {
+    public static void main(String[] args) {
+        Money money = new Money(Integer.parseInt(InputView.inputMoney()));
```
String 값을 Money 생성자로 전달하는 것은 어떨까?

Conatuseus replied:
원래 String값을 생성자로 전달했었는데
저번 리뷰에서 그렇게 하지 말라고 피드백 받았어요.
그래서 다른 입력들에 대해선 parser class를 만들어서 사용했는데 money는 만들지 않았네요.

그럼 MoneyParser을 만들어서 
Money money=MoneyParser.parseMoney(InputView.inputMoney());
이런 식으로 할까 하는데요

String으로 넘겨서 생성자에서 파싱하는 것 vs Parser  클래스를 사용해서 파싱해서 넘겨주는 것
둘 중 어떤게 나은건지 알 수 있을까요?

pobiconan replied:
String, Integer를 모두 받는 두개의 생성자를 추가하는 것은 어떨까요?
어차피 정답은 없으니까 본인이 맞다 생각하는 방향으로 구현하세요.

Conatuseus replied:
네 Parser을 구현해서 넘겨주는 것으로 변경했습니다.
다음에는 말씀해주신 대로 구현해보겠습니다.
아이디어 주셔서 감사합니다!

pobiconan commented on src/test/java/lotto/domain/WinningLottoTest.java:
```diff
@@ -0,0 +1,56 @@
+                        new LottoNumber(3),
+                        new LottoNumber(4),
+                        new LottoNumber(5),
+                        new LottoNumber(6)),new LottoNumber(6)));
```
`new WinningLotto(Arrays.asList(1, 2, 3, 4, 5, 6), 6)`과 같이 테스트 코드를 구현할 수 있는 구조면 어떨까?
API를 사용하는 입장에서 더 좋지 않을까?

Conatuseus replied:
api를 사용하는 입장을 생각하지 않았었네요! 감사합니다.
말씀하신 대로 하자면 Lotto의 생성자에 List<Integer>로 넘겨서
내부에서 Integer를 LottoNumber로 변환하도록 하는게 맞는 것 같아요

pobiconan commented on src/test/java/lotto/domain/UserTest.java:
```diff
@@ -0,0 +1,31 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+public class UserTest {
```
이 테스트는MoneyTest와 중복되는 것 아닌가?

Conatuseus replied:
확인했습니다!

pobiconan commented on src/main/java/lotto/Main.java:
```diff
@@ -0,0 +1,33 @@
+        Money money = new Money(Integer.parseInt(InputView.inputMoney()));
+        int countOfManualLotto = InputView.inputCountOfManualLotto();
+
+        User user = new User(money, countOfManualLotto);
```
굳이 User가 필요한가?
User의 역할은?

pobiconan commented on src/main/java/lotto/Main.java:
```diff
@@ -0,0 +1,33 @@
+        OutputView.printCountOfLotto(money, countOfManualLotto);
+        OutputView.printUserLottos(user.getUserLottos());
+
+        List<LottoNumber> winningLottoNumbers = WinningLottoParser.parseLottoNumbers(InputView.inputWinningLotto());
```
반환 값에 대해서도 일급 콜렉션을 적용해 본다.

Conatuseus replied:
변경했습니다. 저도 그부분에 대해서 고민을 했었는데 생각이 잘 안났어요.
오늘 다시보니 변경할 방법들이 생각나서 고쳤습니다. 감사합니다.

pobiconan commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    boolean isDuplicated(List<LottoNumber> scannedNumbers) {
+        return scannedNumbers.size() != new HashSet<>(scannedNumbers).size();
```
이와 같이 Set으로 변경해 중복을 검증한다면 인스턴스 변수를 `List<LottoNumber> lottoNumbers` 대신 Set을 사용하는 것은 어떨까?

Conatuseus replied:
네 Set으로 수정했습니다.
set으로 가지고 있으면 변경할 필요없이 검증 가능하겠네요!

Conatuseus replied:
Set으로 바꾸고 나니  (중복 검증, 수 크기 검증)을 같이 할 수 있어서 메소드 하나를 줄일 수 있었습니다.
boolean isValidNumbers(Set<LottoNumber> scannedNumbers){
          return scannedNumbers==LOTTO_SIZE; 
}

pobiconan commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,43 @@
+        return scannedNumbers.size() == LOTTO_SIZE;
+    }
+
+    boolean isContains(LottoNumber lottoNumber) {
```
👍

pobiconan commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,43 @@
+        return scannedNumbers.size() != new HashSet<>(scannedNumbers).size();
+    }
+
+    boolean isValidSize(List<LottoNumber> scannedNumbers) {
```
👍

pobiconan commented on src/main/java/lotto/domain/LottoGame.java:
```diff
@@ -0,0 +1,28 @@
+public class LottoGame {
+
+    private final User user;
+    private final Map<Rank, Integer> countOfRank;
```
👍

pobiconan commented on src/main/java/lotto/domain/User.java:
```diff
@@ -0,0 +1,57 @@
+        return this.userLottos;
+    }
+
+    public Map<Rank, Integer> calculateCountOfRank(WinningLotto winningLotto) {
```
이 로직이 WinningLotto에 있는 것은 어떨까?

Conatuseus replied:
수정했습니다.!

pobiconan commented on src/main/java/lotto/domain/User.java:
```diff
@@ -0,0 +1,57 @@
+    private final Money money;
+    private final int countOfLotto;
+    private final int countOfManualLotto;
+    private List<Lotto> userLottos = new ArrayList<>();
```
`List<Lotto> userLottos`에 대한 일급 콜렉션을 만들고 User 객체 없이 구현해 보는 것은 어떨까?
Money의 경우도 사용자가 구매한 로또 목록이 있으면 구할 수 있지 않을까?

pobiconan commented on src/main/java/lotto/domain/UserLotto.java:
```diff
@@ -0,0 +1,31 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class UserLotto extends Lotto {
```
이 객체의 경우 특별히 로직도 없는데 굳이 Lotto를 상속하는 UserLotto를 생성하지 말고 Lotto 자체를 UserLotto 처럼 사용하면 안될까?

Conatuseus replied:
네, toString 메소드를 지우고 나니 로직도 없고 Lotto 와 같은 클래스네요.

pobiconan commented on src/main/java/lotto/domain/UserLotto.java:
```diff
@@ -0,0 +1,31 @@
+
+    private static final String OPEN = "[";
+    private static final String CLOSE = "]";
+    private static final String DELIMITER = ", ";
```
도메인 객체에 View와 관련한 코드를 가지지 않도록 한다.

Conatuseus replied:
수정했습니다!

pobiconan reviewed (changes requested):
전체적인 객체 설계, 단위 테스트 구현 등은 잘했네요. 👍 
앞 단계에서 학습한 내용 잘 반영했는데요. 추가 연습을 위한 피드백 좀 남겼어요.
https://github.com/woowacourse/woowacourse-docs/blob/master/cleancode/pr_checklist.md 체크리스트 참고해 개선할 부분이 있는지 검토해 보고 반영해 보면 좋겠어요.

Conatuseus committed:
refactor : 피드백 반영

Lotto의 생성자 파라미터 타입 변경
parser가 1급객체를 반환하도록 수정
불필요한 userTest 삭제
LottoParser 추가 구현

Conatuseus committed:
refactor : 피드백 반영

반환값 일급 컬렉션 적용
List 대신 Set으로 수정
User 객체 삭제

pobiconan reviewed (approved):
피드백 반영 잘 했네요. 👍
로또 미션의 경우 merge된 후 다음 단계를 진행하려면 woowacouse에 merge된 코드에서 다시 시작해야 충돌이 발생하지 않을 겁니다. 다음과 같은 과정을 진행한 후 다음 단계 진행할 것을 추천해요.
"git remote add upstream" -> "git fetch upstream 자신의 github_id" -> "git reset --hard upstream/자신의 github_id"
웹 미션 기대할께요.

pobiconan commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,44 @@
+package lotto.domain;
+
+
+public class LottoNumber implements Comparable<LottoNumber> {
```
원시 값을 포장한 것은 👍 
단, 이와 같이 포장할 경우 매번 인스턴스를 생성해야 하기 때문에 성능이 떨어진다.
인스턴스를 재사용 가능한 구조로 개선한다.

pobiconan commented on src/test/java/lotto/domain/WinningLottoTest.java:
```diff
@@ -0,0 +1,36 @@
+        assertThrows(InvalidBonusBallException.class, () ->
+                new WinningLotto(Arrays.asList(1,2,3,4,5,6),new LottoNumber(6)));
+    }
+
```
Prize, 1 ~ 5등을 구하는 핵심 로직들에 대한 단위 테스트도 추가하면 좋지 않을까?

pobiconan commented on src/test/java/lotto/domain/WinningLottoTest.java:
```diff
@@ -0,0 +1,36 @@
+    @Test
+    void 보너스볼이_중복된_당첨_로또_생성_검증() {
+        assertThrows(InvalidBonusBallException.class, () ->
+                new WinningLotto(Arrays.asList(1,2,3,4,5,6),new LottoNumber(6)));
```
space convention 위반

pobiconan reviewed (approved):
피드백 반영 잘 했네요. 👍
몇개 피드백 떠 남겼으니 다음 단계 진행할 때 같이 반영해 보세요.
로또 미션의 경우 merge된 후 다음 단계를 진행하려면 woowacouse에 merge된 코드에서 다시 시작해야 충돌이 발생하지 않을 겁니다. 다음과 같은 과정을 진행한 후 다음 단계 진행할 것을 추천해요.
"git remote add upstream" -> "git fetch upstream 자신의 github_id" -> "git reset --hard upstream/자신의 github_id"
웹 미션 기대할께요.
